clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```diff
@@ -495,8 +497,22 @@ public ApiException build(String message) {
             public ApiException build(String message) {
                 return new OperationNotAttemptedException(message);
             }
-        });
-
+    }),
+    KAFKA_STORAGE_ERROR(56, "Disk error when trying to access log file on the disk.",
+        new ApiExceptionBuilder() {
+            @Override
+            public ApiException build(String message) {
+                return new KafkaStorageException(message);
+            }
+    }),
+    UNKNOWN_ERROR_CODE(57, "The client received an unknown error code when processing the response. This may happen if " +
+            "the client library version is lower than the server library version",
+        new ApiExceptionBuilder() {
```

Comment:
Hmm.. not sure if I fully understand the alternative approach.

Let say we introduced two new retriable errors A and B for ProduceResponse in the future. A is added in version 5 and B is added in version 6. If we have UNKNOWN_ERROR_CODE, then the response handling logic would look like this:
```
// Check if the client recognizes UNKNOWN_ERROR_CODE
if (version > 3)
  return original_error_code
else
  return another_existing_error_code (e.g. NOT_LEADER_FOR_PARTITION)
```

On the other hand, if we don't have UNKNOWN_ERROR_CODE, then the response handling logic would look like this, which increases overtime as we have more and more new errors.

```
if (version < 5 and code == A)
  return another_existing_error_code
else if (version < 6 and code == B)
  return another existing_error_code
else
  return original_error_code
```

Thus it seems that UNKNOWN_ERROR_CODE can make things a bit easier. Does this make sense?